agent/consul/controller/controller.go

```diff
@@ -0,0 +1,258 @@
+// queue, this is predominantly just used for testing. This should only ever be called
+// prior to running Start.
+func (c *controller) WithQueueFactory(fn func(ctx context.Context, baseBackoff time.Duration, maxBackoff time.Duration) WorkQueue) Controller {
+	c.makeQueue = fn
+	return c
+}
+
+// Start begins the Controller's main processing loop. When the given
+// context is canceled, the Controller stops processing any remaining work.
+// The Start function should only ever be called once.
+func (c *controller) Start(ctx context.Context) error {
+	group, groupCtx := errgroup.WithContext(ctx)
+
+	// set up our queue
+	c.work = c.makeQueue(groupCtx, c.baseBackoff, c.maxBackoff)
+
+	for _, sub := range c.subscriptions {
+		// store a reference for the closure
+		sub := sub
+		group.Go(func() error {
+			var index uint64
+
+			subscription, err := c.publisher.Subscribe(sub.request)
+			if err != nil {
+				return err
+			}
+			defer subscription.Unsubscribe()
+
+			for {
+				event, err := subscription.Next(ctx)
+				switch {
+				case errors.Is(err, context.Canceled):
+					return nil
+				case err != nil:
+					return err
+				}
+
+				if event.IsFramingEvent() {
+					continue
+				}
+
+				if event.Index <= index {
+					continue
+				}
+
+				index = event.Index
+
+				if err := c.processEvent(sub, event); err != nil {
+					return err
+				}
+			}
+		})
+	}
+
+	for i := 0; i < c.workers; i++ {
+		group.Go(func() error {
+			for {
+				request, shutdown := c.work.Get()
+				if shutdown {
+					// Stop working
+					return nil
+				}
+				c.reconcileHandler(groupCtx, request)
+				// Done is called here because it is required to be called
+				// when we've finished processing each request
+				c.work.Done(request)
+			}
+		})
+	}
+
+	<-groupCtx.Done()
+	return nil
+}
+
+func (c *controller) processEvent(sub subscription, event stream.Event) error {
+	switch payload := event.Payload.(type) {
+	case state.EventPayloadConfigEntry:
+		c.enqueueEntry(sub.transformers, payload.Value)
+		return nil
+	case *stream.PayloadEvents:
+		for _, event := range payload.Items {
+			if err := c.processEvent(sub, event); err != nil {
+				return err
+			}
+		}
+		return nil
+	default:
+		return fmt.Errorf("unhandled event type: %T", payload)
+	}
+}
+
+// enqueueEntry adds all of the given entry into the work queue
+func (c *controller) enqueueEntry(transformers []func(entry structs.ConfigEntry) []Request, entry structs.ConfigEntry) {
+	if len(transformers) == 0 {
+		c.work.Add(Request{
+			Kind: entry.GetKind(),
+			Name: entry.GetName(),
+			Meta: entry.GetEnterpriseMeta(),
+		})
+	} else {
+		for _, fn := range transformers {
+			for _, request := range fn(entry) {
+				c.work.Add(request)
+			}
+		}
+	}
+}
+
+// reconcile wraps the reconciler in a panic handler
+func (c *controller) reconcile(ctx context.Context, req Request) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("panic: %v [recovered]", r)
```

Comment:
#nit I assume `[recovered]` is intended to make it clear that we dealt with it gracefully, so I'd reorder to indicate that as soon as possible instead of on the other end of the (potentially-lengthy) `r`

`panic [recovered]: %v`